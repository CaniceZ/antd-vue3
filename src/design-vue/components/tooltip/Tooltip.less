.ygp-tooltip {
  &.ant-tooltip {
    max-width: 400px;
    color: #fff;
    .ant-tooltip-inner {
      color: inherit;
    }
  }
}
