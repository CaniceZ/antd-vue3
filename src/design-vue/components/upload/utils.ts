export const extname = (url = '') => {
  const temp = url.split('/')
  const filename = temp[temp.length - 1]
  const filenameWithoutSuffix = filename.split(/#|\?/)[0]
  return (/\.[^./\\]*$/.exec(filenameWithoutSuffix) || [''])[0]
}

const isImageFileType = type => !!type && type.indexOf('image/') === 0

export const isImageUrl = file => {
  if (isImageFileType(file.type)) {
    return true
  }
  const url = file.thumbUrl || file.url
  const extension = extname(url)
  if (
    /^data:image\//.test(url) ||
    /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(extension)
  ) {
    return true
  }
  if (/^data:/.test(url)) {
    // other file types of base64
    return false
  }
  if (extension) {
    // other file types which have extension
    return false
  }
  return true
}

export const getFileType = file => {
  if (!file) {
    return
  }
  const attachmentUrl = file.url

  const type = extname(attachmentUrl)
  let fileType = ''
  if (
    isImageFileType(file.type) ||
    /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(type)
  ) {
    fileType = 'image'
  } else if (type === '.pdf') {
    fileType = 'pdf'
  } else if (['.doc', '.dot'].includes(type)) {
    fileType = 'word'
  } else if (['.xls', '.xlsx'].includes(type)) {
    fileType = 'excel'
  } else {
    fileType = 'file'
  }
  return fileType
}

export const imageDimensions = file =>
  new Promise((resolve, reject) => {
    const img = new Image()

    img.onload = () => {
      const { naturalWidth: width, naturalHeight: height } = img
      resolve({ width, height })
    }

    img.onerror = () => {
      reject('There was some problem with the image.')
    }

    img.src = URL.createObjectURL(file)
  })

export const isUrl = (url: string) => {
  const regex = /^(http|https)/
  return regex.test(url.toLowerCase())
}
