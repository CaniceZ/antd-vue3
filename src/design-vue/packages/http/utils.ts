const toString = Object.prototype.toString

export function is(val: unknown, type: string) {
  return toString.call(val) === `[object ${type}]`
}

export function isString(val: unknown): val is string {
  return is(val, 'String')
}

export function isObject(val: any): val is Record<any, any> {
  return val !== null && is(val, 'Object')
}

export function isFunction(val: unknown): val is Function {
  return typeof val === 'function'
}

/**
 * Add the object as a parameter to the URL
 * @param baseUrl url
 * @param obj
 * @returns {string}
 * eg:
 *  let obj = {a: '3', b: '4'}
 *  setObjToUrlParams('www.baidu.com', obj)
 *  ==>www.baidu.com?a=3&b=4
 */
export function setObjToUrlParams(baseUrl: string, obj: any): string {
  let parameters = ''
  for (const key in obj) {
    parameters += key + '=' + encodeURIComponent(obj[key]) + '&'
  }
  parameters = parameters.replace(/&$/, '')
  return /\?$/.test(baseUrl)
    ? baseUrl + parameters
    : baseUrl.replace(/\/?$/, '?') + parameters
}

export function deepMerge<T = any>(src: any = {}, target: any = {}): T {
  let key: string
  for (key in target) {
    src[key] = isObject(src[key])
      ? deepMerge(src[key], target[key])
      : (src[key] = target[key])
  }
  return src
}

declare global {
  interface Navigator {
    msSaveBlob(blob: Blob, string): void
  }
}

/**
 * 根据后端接口文件流下载
 * @param {*} data
 * @param {*} filename
 * @param {*} mime
 * @param {*} bom
 */
export function downloadByData(
  data: BlobPart,
  filename: string,
  mime?: string,
  bom?: BlobPart
) {
  const blobData = typeof bom !== 'undefined' ? [bom, data] : [data]
  const blob = new Blob(blobData, { type: mime || 'application/octet-stream' })
  // 如果浏览器的msSaveBlob存在
  if (typeof window.navigator.msSaveBlob !== 'undefined') {
    // 使用msSaveBlob下载文件
    window.navigator.msSaveBlob(blob, filename)
  } else {
    // 创建一个URL
    const blobURL = window.URL.createObjectURL(blob)
    const tempLink = document.createElement('a')
    tempLink.style.display = 'none'
    tempLink.href = blobURL
    tempLink.setAttribute('download', filename)
    if (typeof tempLink.download === 'undefined') {
      tempLink.setAttribute('target', '_blank')
    }
    document.body.appendChild(tempLink)
    // 手动触发点击事件
    tempLink.click()
    // 点击完成后移除链接
    document.body.removeChild(tempLink)
    // 移出URL
    window.URL.revokeObjectURL(blobURL)
  }
}
