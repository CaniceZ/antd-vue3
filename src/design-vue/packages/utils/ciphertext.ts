import get from 'lodash-es/get'

type Options = {
  sign: '*' | string
  suffix: '_encrypt' | string
}

const defaultOptions: Options = {
  sign: '*',
  suffix: '_encrypt'
}

/**
 * 获取密文的密钥字段名
 * @param name
 * @param opts
 * @returns
 */
export function getCiphertextField(
  name: string,
  opts: Partial<Options> = defaultOptions
): string {
  const options = {
    ...defaultOptions,
    ...opts
  }
  return `${name}${options.suffix}`
}

/**
 * 获取密文对象实例
 * @param model 数据对象
 * @param name 密文字段名称 / 密文字段路径
 * @param opts.sign 脱敏显示符号
 * @param opts.suffix 脱敏密文字段名后缀
 * @returns
 */
export function getCiphertext(
  model: object,
  name: string,
  opts: Partial<Options> = defaultOptions
): {
  text: string
  cipherText: string
} {
  const options = {
    ...defaultOptions,
    ...opts
  }
  const emptyRes = {
    text: '',
    cipherText: ''
  }
  if (typeof model === 'object') {
    const text = get(model, name)
    const cipherText = get(model, getCiphertextField(name, options))
    if (text && cipherText && text.includes(options.sign)) {
      return {
        text,
        cipherText
      }
    } else {
      return emptyRes
    }
  } else {
    return emptyRes
  }
}

/**
 * 判断是否密文
 * @param model 数据对象
 * @param name 密文字段名称 / 密文字段路径
 * @param opts.sign 脱敏显示符号
 * @param opts.suffix 脱敏密文字段名后缀
 */
export function isCiphertext(
  model: object,
  name: string,
  opts: Partial<Options> = defaultOptions
): boolean {
  const options = {
    ...defaultOptions,
    ...opts
  }
  const ciphertext = getCiphertext(model, name, options)
  return Boolean(ciphertext.text && ciphertext.cipherText)
}
