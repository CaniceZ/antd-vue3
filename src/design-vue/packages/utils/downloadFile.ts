/**
 * 下载文件
 * @param url 下载路径
 * @returns void
 */
export default function downloadFile(url: string) {
  if (!url) return
  const a = document.createElement('a')
  a.href = url
  a.click()
}
