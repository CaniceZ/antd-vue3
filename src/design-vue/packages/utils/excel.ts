import { Message, Modal, YImportResultModal } from '../../'

/**
 * 导入文件
 * @param params
 * @param callback
 * @returns Promise
 */
export const importFile = async (params: {
  file: File | any
  title: string
  getBatchNumber: Function
  getImportResult: Function
  wrapCount?: number
  extraData?: Object
  templateCode?: string
  resultProps?: any
  resultSlots?: any
}) => {
  const {
    file,
    getBatchNumber,
    getImportResult,
    title,
    extraData,
    templateCode,
    resultProps,
    resultSlots,
    wrapCount
  } = params

  const formData = new FormData()
  formData.append('file', file)
  templateCode && formData.append('templateCode', templateCode) // 模板编码

  if (extraData) {
    Object.keys(extraData).forEach(key => {
      formData.append(key, extraData![key])
    })
  }

  const res: any = await getBatchNumber(formData)
  return new Promise((resolve, reject) => {
    const batchNumber = res?.batchNumber || res
    if (batchNumber) {
      Message.success('上传成功，稍后通知您导入结果!')
      importResultPolling({
        resultProps,
        resultSlots,
        getImportResult,
        batchNumber,
        title,
        wrapCount,
        doneCallback: () => {
          resolve(params)
        },
        timeoutCallback: () => {
          reject()
        }
      })
    } else {
      Message.error('上传失败，请稍后重试')
      reject()
    }
  })
}

/**
 * 导入结果轮询
 * @param params 轮询的请求参数
 * @param callback 成功回调
 */
export const importResultPolling = (params: {
  title: String
  batchNumber: string
  getImportResult: Function
  resultProps?: any
  resultSlots?: any
  wrapCount?: number
  doneCallback?: (...args: any) => void
  timeoutCallback?: (...args: any) => void
}) => {
  const {
    wrapCount,
    batchNumber,
    title,
    doneCallback,
    timeoutCallback,
    resultProps,
    resultSlots,
    getImportResult
  } = params
  if (!batchNumber) return
  function run(count = 1) {
    setTimeout(() => {
      let params = { batchNumber }
      if (resultProps?.extraData) {
        params = { ...params, ...resultProps.extraData }
      }
      const typeMap = resultProps?.typeMap
      getImportResult(params).then(
        res => {
          if (resultProps?.isPms) {
            // pms特殊导入结果处理
            if (res && res.id) {
              if (!resultProps?.hideModal) {
                Modal.create(YImportResultModal, {
                  list: [res],
                  typeMap,
                  importBizSet: [resultProps.importBiz],
                  title,
                  ...(resultProps || {})
                })
              }
              doneCallback?.({ res, typeMap, title })
            } else {
              if (count < (wrapCount || 20)) {
                run(++count)
              } else {
                Message.error('稍后请在最近导入结果中查看')
                timeoutCallback?.()
              }
            }
          } else {
            // excel通用
            const { status, msg } = res
            if (status === 4) {
              if (count < (wrapCount || 20)) {
                run(++count)
              } else {
                Message.error(msg || '稍后请在最近导入结果中查看')
                timeoutCallback?.()
              }
            } else {
              if (resultProps?.callback) {
                resultProps.callback(res)
              } else {
                if (!resultProps?.hideModal) {
                  Modal.create(
                    YImportResultModal,
                    {
                      title,
                      list: [res],
                      ...(resultProps || {})
                    },
                    resultSlots || {}
                  )
                }
                doneCallback?.({ res, typeMap, title })
              }
            }
          }
        },
        err => {
          Message.error(err?.msg || '上传失败，请稍后重试')
        }
      )
    }, 3000)
  }
  run()
}

// pms导出结果轮询
export const pmsExportPolling = (option: {
  batchNumber: string
  exportBiz: String
  userCode: String
  getPmsExportDetail: Function
  timeout?: (...args: any) => void
  done?: (...args: any) => void
  error?: (...args: any) => void
}) => {
  const {
    batchNumber,
    exportBiz,
    userCode,
    getPmsExportDetail,
    timeout,
    done,
    error
  } = option
  if (!batchNumber) return
  const params = {
    batchNumber: batchNumber,
    exportBiz,
    userCode
  }
  function run(count = 1) {
    setTimeout(() => {
      getPmsExportDetail(params).then(
        res => {
          const { code, data, msg } = res
          if (code === 0) {
            if (data && data.list.length > 0) {
              done?.(data)
              data.list &&
                data.list.forEach(item => {
                  if (item.excelUrl) {
                    window.location.href = item.excelUrl
                  } else {
                    Message.error(item.msg)
                  }
                })
            } else {
              if (count < 40) {
                run(++count)
              } else {
                Message.error('导出超时，请重新导出')
                timeout?.()
              }
            }
          } else {
            Message.error(msg)
            error?.()
          }
        },
        () => {
          error?.()
        }
      )
    }, 3000)
  }
  run()
}
