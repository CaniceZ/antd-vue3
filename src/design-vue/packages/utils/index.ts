import omit from './omit'
import pick from './pick'
import downloadFile from './downloadFile'
import { isCiphertext, getCiphertext, getCiphertextField } from './ciphertext'
import { needCLodop, loadCLodop, getLodop } from './LodopFuncs'
import { importFile, importResultPolling } from './excel'
import filterUnit from './filterUnit'
import filterMoney from './filterMoney'

export {
  omit,
  pick,
  downloadFile,
  isCiphertext,
  getCiphertext,
  getCiphertextField,
  needCLodop,
  loadCLodop,
  getLodop,
  importFile,
  importResultPolling,
  filterUnit,
  filterMoney
}
