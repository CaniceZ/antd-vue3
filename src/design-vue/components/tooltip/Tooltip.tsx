import { omit } from '../../packages//utils'
import {
  Tooltip as ATooltip,
  TooltipProps as ATooltipProps
} from 'ant-design-vue'
import { defineComponent, onMounted, PropType, useSlots, watch } from 'vue'
import { GlobalComponentConstructor, PropsType } from '../../types/ts-helpers'
// import useActualProps from '../../_hooks/useActualProps'
import useAttrs from '../../_hooks/useAttrs'
import { getPrefix } from '../../_utils/common'
import { ATooltipSlots } from './type'

const [prefixName, prefixCls] = getPrefix('tooltip')

const tooltipProps = () => ({
  modelValue: {
    type: Boolean as PropType<ATooltipProps['visible']>,
    default: false
  },
  textColor: {
    type: String
  }
})

const Tooltip = defineComponent({
  name: prefixName,
  inheritAttrs: false,
  props: {
    ...tooltipProps()
  },
  emits: {
    'update:modelValue': (_visible: ATooltipProps['visible']) => true
  },
  setup(props, { emit, expose }) {
    const attrs = $(useAttrs<ATooltipProps>(['defaultVisible']))
    const slots = useSlots() as ATooltipSlots
    // const actualProps = useActualProps<PropsType<typeof Tooltip>>()

    let visible = $ref<ATooltipProps['visible']>(props.modelValue)
    watch($$(visible), val => {
      emit('update:modelValue', val)
    })
    watch(
      () => props.modelValue,
      val => {
        visible = val
      }
    )

    onMounted(() => {
      if (attrs.defaultVisible) {
        visible = attrs.defaultVisible
      }
    })

    function hide() {
      visible = false
    }
    function show() {
      visible = true
    }

    expose({ hide, show })

    return () => (
      <ATooltip
        destroyTooltipOnHide={true}
        {...omit(attrs, ['visible', 'onUpdate:visible'])}
        overlayClassName={`${prefixCls} ${attrs.overlayClassName || ''}`}
        overlayStyle={{
          color: props.textColor,
          ...(attrs.overlayStyle || {})
        }}
        visible={visible}
        onUpdate:visible={val => {
          if (props?.modelValue) {
            visible = props.modelValue
          } else {
            visible = val
          }
        }}
        v-slots={slots}
      />
    )
  }
})

export type TooltipProps = PropsType<
  Omit<ATooltipProps, 'visible' | 'onUpdate:visible'>
> &
  PropsType<typeof Tooltip> &
  PropsType<typeof tooltipProps>
export type TooltipExpose = {
  show?: () => void
  hide?: () => void
}
export type TooltipSlots = ATooltipSlots

export default Tooltip as GlobalComponentConstructor<TooltipProps, TooltipSlots>
