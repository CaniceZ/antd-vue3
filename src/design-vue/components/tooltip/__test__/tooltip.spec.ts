import { mount } from '@vue/test-utils'
import { asyncExpect } from '../../../tests/utils'
import Tooltip from '..'
import Ygpd from '../../../index'

describe('YTooltip', () => {
  // const { appendChild } = document

  it('YTooltip render is ok?', async () => {
    const wrapper: any = mount(Tooltip as any, {
      props: {
        title: 'title="tooltip一直显示',
        modelValue: true
      },

      global: {
        plugins: [Ygpd]
      },
      slots: {
        default: '<span>tooltip一直显示</span>'
      }
    })
    // wrapper.appendChild = appendChild

    // wrapper.setProps({
    //   getPopupContainer: () => {
    //     return wrapper.find('.ant-tooltip-open')
    //   },
    // })

    // 有这个class则证明ant的tooltip渲染成功
    expect(wrapper.find('.ant-tooltip-open').exists()).toBe(true)
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot()
    }, 0)
  })
})
