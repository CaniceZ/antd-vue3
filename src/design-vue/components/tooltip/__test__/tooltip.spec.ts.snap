// Vitest Snapshot v1

exports[`YTooltip > YTooltip render is ok? 1`] = `
"<!--teleport start-->
<!--teleport end-->
<span class=\\"ant-tooltip-open\\">tooltip一直显示</span>"
`;
